### ISSUE TITLE
Add retry for S3 rate-limit failures

### ISSUE THREAD

**Issue by cjames53** [pgbackrest/pgbackrest#622]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.06

2. PostgreSQL version: 9.6

3. Operating system/version: Ubuntu 16.04

4. Did you install pgBackRest from source or from a package? package

5. Please attach the following as applicable:

```
[global]
  backup-user=postgres
[emolecules]
  pg1-path=/var/lib/postgresql/9.6/main
  backup-standby=y
  process-max=4
  repo-type=s3
  repo-cipher-type=none
  repo-path=/
  repo1-retention-diff=7
  repo1-retention-full=3
  repo-s3-bucket=xxxxxxx
  repo-s3-endpoint=s3.amazonaws.com
  repo-s3-key=xxxx
  repo-s3-key-secret=xxxxxx
  repo-s3-region=us-west-2
  start-fast=y
  stop-auto=y
[global:archive-push]
  compress-level=3
---------------------

listen_addresses = '*'
max_connections = 2000
wal_level = logical
wal_keep_segments = 1000
max_wal_size = 4GB
max_wal_senders = 10
hot_standby = on
archive_mode = on
archive_command = 'pgbackrest --stanza=emolecules archive-push %p'
max_replication_slots = 10
```

7. Describe the issue:

S3 archiving gets very occasional random failures. The problem is, our database has LOTS of tables (>200,000), so "occasional" means a large fraction of archive operations fail, like more than 50% of the time. This happens on both backup and restore. We had no such problems when using NFS.

For example, we were at the 100% mark of a restore, with over 600,000 files restored in six hours, and one of the empty (zero byte) files failed. Now we have to start over (fortunately it was for testing, not production).

It would be very helpful to provide a retry feature, maybe with a timer that backs off more with each try, e.g. try again immediately, then after 5, 10, 30, and 60 seconds, and only give up if all the retries fail. I believe these S3 errors are transitory and usually won't happen twice in a row.

Here is the error we get:
```
2018-10-30 00:48:01.674 P00  ERROR: [039]: raised from local-1 process: S3 request error [503] Service
                                        *** request header ***
                                        GET /backup/emolecules/20181028-135420F/pg_data/base/633488/47771218.gz? HTTP/1.1
                                        authorization: AWS4-HMAC-SHA256 Credential=xxxxx/xxxxx/us-west-2/s3/aws4_request,SignedHeaders=content-length;host;x-amz-content-sha256;x-amz-date,Signature=xxxxx
                                        content-length: 0
                                        host: emol-db-backup-prod-us-west-2.s3.amazonaws.com
                                        x-amz-content-sha256: xxxxx
                                        x-amz-date: 20181030T004801Z
                                        *** response header ***
                                        Date: Tue, 30 Oct 18 00:48:01 GMT
                                        Connection: close
                                        x-amz-id-2: xxxxx
                                        x-amz-request-id: xxxxxxxx
                                        Content-Type: application/xml
                                        Server: AmazonS3
                                        Content-Length: 0
2018-10-30 00:48:01.675 P00  ERROR: [039]: raised from local-1 process: S3 request error [503] Service
                                        *** request header ***
                                        GET /backup/emolecules/20181028-135420F/pg_data/base/633488/47771218.gz? HTTP/1.1
                                        authorization: AWS4-HMAC-SHA256 Credential=xxxxx/xxxxx/us-west-2/s3/aws4_request,SignedHeaders=content-length;host;x-amz-content-sha256;x-amz-date,Signature=xxxxx
                                        content-length: 0
                                        host: emol-db-backup-prod-us-west-2.s3.amazonaws.com
                                        x-amz-content-sha256: xxxxx
                                        x-amz-date: 20181030T004801Z
                                        *** response header ***
                                        Date: Tue, 30 Oct 18 00:48:01 GMT
                                        Connection: close
                                        x-amz-id-2: xxxxx
                                        x-amz-request-id: xxxxx
                                        Content-Type: application/xml
                                        Server: AmazonS3
                                        Content-Length: 0
2018-10-30 00:48:02.256 P00   INFO: restore command end: aborted with exception [039]
```

**Comment by dwsteele** [comment]
We definitely do retry -- even a trivial restore generally won't succeed without retries.

But -- we are not currently retrying on 503 which seem to be a rate-limiting mechanism.  What are you using for `process-max`?  Looking more closely at the docs it seems we should be retrying on any error in the 500 range so we'll add that.

BTW, it's not ideal, but if you add `--delta` to your restores at least you won't need to redo the files that have already been restored.  You are probably already doing this but I thought I would mention it.

**Comment by dwsteele** [comment]
This has been improved in cca7a4ffd477.  Let me know what you think.

**Comment by cjames53** [comment]
Thanks! Glancing at the code changes it all looks good. I'll test it as soon as I can, a couple days at most.

**Comment by dwsteele** [comment]
This was released in 2.07.
